### PR TITLE
fabtests/unexpected_msg: add testing for FI_ADDR_UNSPEC, modify defaults for tcp

### PR DESCRIPTION
**Pull request by aingerson**

### CODE REVIEW

**Comment by shijin-aws** [comment]
The AWS CI failure are expected, because efa will not pass this test

File: fabtests/functional/unexpected_msg.c [hunk 1]
```diff
@@ -199,6 +200,96 @@ static int run_test_loop(void)
...
 }
 
+static int exchange_unexp_addr(void)
+{
+	char temp[FT_MAX_CTRL_MSG];
+	size_t addrlen;
+	int ret;
+
+	if (opts.dst_addr) {
+		ret = ft_av_insert(av, fi->dest_addr, 1, &remote_fi_addr, 0,
```
Comment (shijin-aws) [inline fabtests/functional/unexpected_msg.c:210]: This line caused the segfault for efa provider. Efa provider currently only support out-of-band address exchange in ft_init_av, where the address to be inserted is the ep name sent from peer via sockets.

What is the fi->dest_addr here in this case?

File: fabtests/functional/unexpected_msg.c [hunk 2]
```diff
@@ -199,6 +200,96 @@ static int run_test_loop(void)
...
+		 */
+		ret = ft_post_tx_buf(ep, remote_fi_addr, addrlen, 0, &tx_ctx,
+				     tx_buf, mr_desc, ft_tag);
+		if (ret)
+			goto err;
+
+		ret = ft_post_tx_buf(ep, remote_fi_addr, addrlen, 0, &tx_ctx,
+				     tx_buf, mr_desc, ft_tag);
+		if (ret)
+			return ret;
```
Comment (shijin-aws) [inline fabtests/functional/unexpected_msg.c:238]: do you need to get_tx_comp here?

File: fabtests/functional/unexpected_msg.c [hunk 3]
```diff
@@ -199,6 +200,96 @@ static int run_test_loop(void)
...
+			goto err;
+
+		ret = ft_get_rx_comp(rx_seq);
+		if (ret)
+			goto err;
+	} else {
+		ft_sync();
+
+		/* Process first unexpected message with unspec addr*/
+		ret = ft_post_rx(ep, rx_size, &rx_ctx);
```
Comment (shijin-aws) [inline fabtests/functional/unexpected_msg.c:254]: there was a rx posted in ft_ep_enable_recv. Do we need this ?

File: fabtests/functional/unexpected_msg.c (hunk 2 above)
Comment (aingerson) [reply fabtests/functional/unexpected_msg.c:238]: yes you're right, thanks!

File: fabtests/functional/unexpected_msg.c (hunk 3 above)
Comment (aingerson) [reply fabtests/functional/unexpected_msg.c:254]: the pre posted receive only happens if we use the common code message allocation which is avoided in this test by setting
FT_OPT_SKIP_MSG_ALLOC

Comment (shijin-aws) [reply fabtests/functional/unexpected_msg.c:254]: I see, I missed that we have a separate buf in this test.

File: fabtests/functional/unexpected_msg.c [hunk 4]
```diff
@@ -223,7 +319,9 @@ int main(int argc, char **argv)
 
 	opts = INIT_OPTS;
 	opts.iterations = 600; // Change default from 1000.
-	opts.options |= FT_OPT_OOB_CTRL | FT_OPT_SKIP_MSG_ALLOC;
+	opts.transfer_size = 128;
+	opts.options |= FT_OPT_OOB_CTRL | FT_OPT_SKIP_MSG_ALLOC |
+		        FT_OPT_SKIP_ADDR_EXCH;
```
Comment (shijin-aws) [inline fabtests/functional/unexpected_msg.c:324]: I didn't realize we have such option to skip addr exchange. TIL.

**Comment by aingerson** [comment]
@shijin-aws Updated to include proper support for the OOB address exchange. Let me know if efa still has issues or behaves as expected now.

**Comment by shijin-aws** [comment]
Thx, I think it will fail for efa on single node. I will have a PR up to temporarily skip it.

**Comment by shijin-aws** [comment]
In AWS CI, shm provider failed as

```
server_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.19.31 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr8914/install/fabtests/bin/fi_unexpected_msg -e rdm -I 10 -p shm -s 172.31.19.31'"'"''

client_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.19.31 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr8914/install/fabtests/bin/fi_unexpected_msg -e rdm -I 10 -p shm -s 172.31.19.31 172.31.19.31'"'"''
client_stdout:
unexpected address exchange error(): functional/unexpected_msg.c:286, ret=-257 (Provided buffer is too small)

client returncode: 1
server_stdout:
unexpected address exchange error(): functional/unexpected_msg.c:286, ret=-257 (Provided buffer is too small)

server returncode: 1
```

Is that expected?

**Comment by shijin-aws** [comment]
tcp provider failed in both single-node and multi-node.

Single node failure looks similar to shm's. multi-node failed as

```
server_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.8.136 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr8914-debug/install/fabtests/bin/fi_unexpected_msg -e rdm -I 10 -p tcp -s 172.31.8.136'"'"''

client_command: ssh -n -o StrictHostKeyChecking=no -o ConnectTimeout=30 -o BatchMode=yes 172.31.8.136 'timeout 360 /bin/bash --login -c '"'"'/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr8914-debug/install/fabtests/bin/fi_unexpected_msg -e rdm -I 10 -p tcp -s 172.31.8.136 172.31.8.136'"'"''
client_stdout:
[error] fabtests:common/shared.c:2665: cq_readerr 125 (Operation canceled), provider errno: 11 (Resource temporarily unavailable)
unexpected address exchange error(): functional/unexpected_msg.c:286, ret=-125 (Operation canceled)

client returncode: 125
server_stdout:
[error] fabtests:common/shared.c:2665: cq_readerr 265 (Truncation error), provider errno: 11 (Resource temporarily unavailable)
unexpected address exchange error(): functional/unexpected_msg.c:286, ret=-265 (Truncation error)

server returncode: 1
```

**Comment by aingerson** [comment]
@shijin-aws I had an uninitialized variable. Just fixed and pushed. Hopefully that fixes the issue!

**Comment by shijin-aws** [comment]
@aingerson tests looks cleaner for the new revision, only the single node efa failed which is expected. I opened a PR https://github.com/ofiwg/libfabric/pull/8935 to temporarily skip this test. Would u mind rebasing on it after it's merged so we can get the CI passed?

File: fabtests/functional/unexpected_msg.c [hunk 5]
```diff
@@ -318,6 +318,7 @@ int main(int argc, char **argv)
 
 	opts = INIT_OPTS;
 	opts.iterations = 600; // Change default from 1000.
+	opts.transfer_size = 128;
```
Comment (shefty) [inline fabtests/functional/unexpected_msg.c:321]: Note that PR #8920 removes the tcp restrictions

**Review by shefty — APPROVED** [review]

File: fabtests/functional/unexpected_msg.c (hunk 5 above)
Comment (aingerson) [reply fabtests/functional/unexpected_msg.c:321]: Do you want me to remove the changes here and wait for #8920? I already have to rebase to pick up the EFA changes in #8935

Comment (shefty) [reply fabtests/functional/unexpected_msg.c:321]: No need -- we can keep these for now.  I have other changes to this test out of tree that I'll eventually push upstream after these are merged.

**Comment by shijin-aws** [comment]
I merged https://github.com/ofiwg/libfabric/pull/8935, please rebase.

**Review by shijin-aws — APPROVED** [review]
AWS CI passed. Ready for merge.

**Pull request merged**